exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,398 @@
+        Err(TransactionError::Code(c)) => u16::from(c),
+        Err(TransactionError::Description(_)) => TRANSACTION_STATUS_DESCRIPTION,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use futures::sync::mpsc;
+
+    use std::collections::BTreeMap;
+    use std::sync::Mutex;
+
+    use super::*;
+    use crypto;
+    use blockchain::Blockchain;
+    use storage::{Database, MemoryDB, Entry};
+    use node::ApiSender;
+    use helpers::{ValidatorId, Height};
+
+    lazy_static! {
+        static ref EXECUTION_STATUS: Mutex<ExecutionResult> = Mutex::new(Ok(()));
+    }
+
+    #[test]
+    fn transaction_status_round_trip() {
+        let statuses = [
+            Ok(()),
+            Err(TransactionError::Panic),
+            Err(TransactionError::Failure),
+            Err(TransactionError::Code(0)),
+            Err(TransactionError::Code(1)),
+            Err(TransactionError::Code(100)),
+            Err(TransactionError::Code(254)),
+            Err(TransactionError::Code(255)),
+            Err(TransactionError::Description("".to_owned())),
+            Err(TransactionError::Description("e".to_owned())),
+            Err(TransactionError::Description("just error".to_owned())),
+            Err(TransactionError::Description(
+                "(Not) really long error description".to_owned(),
+            )),
+            Err(TransactionError::Description(
+                "_underscored_text_".to_owned(),
+            )),
+            Err(TransactionError::Description("!@#$%^&*()".to_owned())),
+        ];
+
+        for status in &statuses {
+            let bytes = status.clone().into_bytes();
+            let new_status = TransactionResult::from_bytes(Cow::Borrowed(&bytes));
+            assert_eq!(*status, new_status);
+        }
+    }
+
+    #[test]
+    fn error_discards_transaction_changes() {
+        let statuses = [
+            Err(ExecutionError::Failure),
+            Err(ExecutionError::Code(0)),
+            Err(ExecutionError::Code(255)),
+            Err(ExecutionError::Description("".to_owned())),
+            Err(ExecutionError::Description("Error description".to_owned())),
+            Ok(()),
+        ];
+
+        let (_, sec_key) = crypto::gen_keypair();
+        let (blockchain, mut pool) = create_blockchain();
+        let db = Box::new(MemoryDB::new());
+
+        for (index, status) in statuses.iter().enumerate() {
+            let index = index as u64;
+
+            *EXECUTION_STATUS.lock().unwrap() = status.clone();
+
+            let transaction = TxResult::new(index, &sec_key);
+            pool.insert(
+                transaction.hash(),
+                Box::new(transaction.clone()) as Box<Transaction>,
+            );
+
+            let (_, patch) = blockchain.create_patch(
+                ValidatorId::zero(),
+                Height(index),
+                &[transaction.hash()],
+                &pool,
+            );
+
+            db.merge(patch).unwrap();
+
+            let mut fork = db.fork();
```

Comment:
Nitpick: Using a `Fork` here looks a bit sloppy, but I suppose it's OK for tests.

Reply:
Yeap, it is only needed to reuse the `create_entry` function.